flight/Libraries/uavtalk.c

```diff
@@ -617,6 +640,95 @@ static int32_t sendBuf(UAVTalkConnection connectionHandle, uint8_t *buf, uint16_
 }
 
 /**
+ * Handles a request for file data.
+ * \param[in] connection The connection on which a request was just received.
+ */
+static void handleFileReq(UAVTalkConnectionData *connection)
+{
+	UAVTalkInputProcessor *iproc = &connection->iproc;
+	uint32_t file_id = iproc->objId;
+
+	struct filereq_data *req = (struct filereq_data *) connection->rxBuffer;
+
+	/* printf("Got filereq for file_id=%08x offs=%d\n", file_id, req->offset); */
+
+	/* Need txbuffer, and need to make sure file response msgs
+	 * are contiguous on link.
+	 */
+	PIOS_Recursive_Mutex_Lock(connection->lock, PIOS_MUTEX_TIMEOUT_MAX);
+
+	connection->txBuffer[0] = UAVTALK_SYNC_VAL;  // sync byte
+	connection->txBuffer[1] = UAVTALK_TYPE_FILEDATA;
+	// data length inserted here below
+	connection->txBuffer[4] = (uint8_t)(file_id & 0xFF);
+	connection->txBuffer[5] = (uint8_t)((file_id >> 8) & 0xFF);
+	connection->txBuffer[6] = (uint8_t)((file_id >> 16) & 0xFF);
+	connection->txBuffer[7] = (uint8_t)((file_id >> 24) & 0xFF);
+
+	uint8_t data_offs = 8;
+
+	struct fileresp_data *resp =
+		(struct fileresp_data *) (connection->txBuffer + data_offs);
+
+	data_offs += sizeof(*resp);
+
+	uint32_t file_offset = req->offset;
+
+	for (int i = 0; ; i++) {
+		resp->offset = file_offset;
+		resp->flags = 0;
+
+		int32_t cb_numbytes = -1;
+
+		if (connection->fileCb) {
+			cb_numbytes = connection->fileCb(connection->cbCtx,
+				connection->txBuffer + data_offs,
+				file_id, file_offset, 100);
+		}
+
+		uint8_t total_len = data_offs;
+
+		if (cb_numbytes > 0) {
+			total_len += cb_numbytes;
+
+			file_offset += cb_numbytes;
+
+			if (i == 5) {	/* 6 messages per chunk */
+				resp->flags = UAVTALK_FILEDATA_LAST;
+			} else {
+				resp->flags = 0;
+			}
+		} else {
```

Comment:
So if no callback registered or an error getting the file, 0 bytes response with EOF.. seems reasonable. ENOENT could also be nice to have?